### DISCUSSION TITLE
InputText label and multiple character deletes

### DISCUSSION THREAD

**Discussion by bs-lab** [Gnimuc/CImGui.jl#73] [Q&A] [answered]
First of all, I just started to use this package, and I'm really digging it.  

The problem I'm having is with InputText. If I select multiple characters from the InputText (either by using the mouse or with shift + arrow keys) and then delete the selection (using backspace or delete), the GUI will properly show the remaining string, but internally, the character array is only ever shorted by one.

So in the following, if the user highlights and then deletes "abc" from the original "abcdefghi" string, the GUI shows "defghi", but the value of the user_text variable becomes "defghihi".

    @cstatic user_text="abcdefghi\0" begin
        CImGui.InputText("##usertextbox", user_text, length(user_text))
        println("user_text:  ", user_text)
    end

As a workaround, is there maybe some way that I can extract the text that's being correctly rendered on screen?

**Comment by Gnimuc** [comment, marked as answer]
That's just an example. In a real-world project, you should use structs for state management. Basically, you could use a char array as a container for storing the string and an integer for recording the length of the current string.

**Reply by bs-lab** [reply]
Thanks! Switching to using a struct led me to see that the char array had a NULL character in the correct spot when printed out. In other words, "defghihi" was really "defghi\0hi\0". That was enough to get me a working solution by replacing the stale remaining characters in the array with more NULLs. I suspect that I've overly complicated things though, or maybe that's just how it is when working closely with C++.

I defined a struct outside of the rendering loop:

    mutable struct State
        user_value::String
        max_len::Int64     # includes terminating NULL character
        curr_len::Int64    # excludes terminating NULL character
    end
                                                                                                          
    my_state = State("abcdefghi\0", 10, 9)

And in the loop:

    CImGui.InputText("## enter value", my_state.user_value, my_state.max_len)
    value, _ = split(my_state.user_value, '\0', limit=2)
    my_state.user_value = value * '\0'^(my_state.max_len - length(value))
    my_state.curr_len = length(value)
    println("user_text: ", my_state.user_value)
